File: backend/middlewares/auth.middleware.js
import { verifyAccessToken } from "../utils/jwt.js";

export function requireAuth(req, res, next) {
  const token = req.cookies?.access_token;
  if (!token) {
    return res.status(401).json({ error: "Not authenticated" });
  }
  try {
    const payload = verifyAccessToken(token);
    req.userId = payload.sub;
    next();
  } catch {
    return res.status(401).json({ error: "Invalid or expired access token" });
  }
}

File: backend/middlewares/auth_middleware.py
from fastapi import Depends, Request

from lib.prisma_client import db
from middlewares.error_handler import AuthError
from utils.jwt_utils import verify_access_token


async def require_auth(request: Request) -> str:
    """FastAPI dependency equivalent of requireAuth — use as:
    `user_id: str = Depends(require_auth)` on a route.
    Returns the userId (payload["sub"]), same as req.userId did in Express."""
    token = request.cookies.get("access_token")
    if not token:
        raise AuthError("Not authenticated")
    try:
        payload = verify_access_token(token)
    except Exception:
        raise AuthError("Invalid or expired access token")
    return payload["sub"]


async def require_admin(user_id: str = Depends(require_auth)) -> str:
    """Admin-only dependency — use as `user_id: str = Depends(require_admin)`.
    Checks role from DB (not the JWT) so a role downgrade takes effect immediately
    instead of waiting for the access token to expire."""
    user = await db.user.find_unique(where={"id": user_id})
    if not user or user.role != "ADMIN":
        raise AuthError("Admin access required", 403)
    return user_id
